allennlp/data/data_loaders/multi_process_data_loader.py

```diff
@@ -0,0 +1,304 @@
+
+        for instances_chunk in lazy_groups_of(instances, self._INSTANCE_CHUNK_SIZE):
+            queue.put(instances_chunk)
+
+        # Indicate to the consumer that this worker is finished.
+        queue.put(None)
+
+        # Wait for consumer to finish to avoid prematurely closing the queue.
+        queue.join()
+
+    def _instances_to_batches(self, instance_iterator: Iterable[Instance]) -> Iterator[TensorDict]:
+        instance_chunks: Iterable[List[Instance]]
+        if self.max_batches_in_memory is not None:
+            chunk_size = self.batch_size * self.max_batches_in_memory
+            instance_chunks = lazy_groups_of(instance_iterator, chunk_size)
+        else:
+            instance_chunks = [list(instance_iterator)]
+
+        for instances in instance_chunks:
+            if self.shuffle:
+                random.shuffle(instances)
+
+            batches: Iterator[List[Instance]]
+            if self.batch_sampler:
+                batches = (
+                    [instances[i] for i in batch_indices]
+                    for batch_indices in self.batch_sampler.get_batch_indices(instances)
+                )
+            else:
+                batches = lazy_groups_of(instances, self.batch_size)
+
+            batched_tensor_dicts = (self.collate_fn(batch) for batch in batches)
+
+            yield from batched_tensor_dicts
+
+    def _iter_batches(self) -> Iterator[TensorDict]:
+        if self._instances is not None or self.num_workers <= 0:
+            for batch in self._instances_to_batches(self.iter_instances()):
+                yield batch
+        else:
+            # At this point self.max_batches_in_memory is not None since lazy must be False.
+            assert self.max_batches_in_memory is not None
+
+            # First we start "instance workers", which are in charge generating raw
+            # instances using self.reader. The generated instances are then put
+            # into the `instance_queue` for the `batch_worker` to consume.
+            instance_queue: mp.JoinableQueue = mp.JoinableQueue(self._INSTANCE_QUEUE_SIZE)
+            instance_workers = self._start_instance_workers(instance_queue)
+
+            # Now start the `batch_worker`. This worker consumes from the `instance_queue`
+            # and puts the resulting batches into the `batch_queue`.
+            batch_queue: mp.JoinableQueue = mp.JoinableQueue(self.max_batches_in_memory)  # type: ignore
+            batch_worker = mp.Process(
+                target=self._batch_worker, args=(instance_queue, batch_queue,), daemon=True
+            )
+            batch_worker.start()
+
+            # We can now start consuming from the `batch_queue` as the `batch_worker`
+            # produces batches.
+            batch_group: Optional[List[TensorDict]]
+            for batch_group in iter(batch_queue.get, None):
+                yield from batch_group
+                batch_queue.task_done()
```

Comment:
If all we're doing here is read from the batch queue, why can't this process, the parent process, be the `batch_worker`?

Reply:
For non-lazy loading, you're right, we don't gain anything from having `_batch_worker` run in another process. But for lazy loading this is important since `_batch_worker` handles the batch sampling. Doing the batch sampling in the main process would slow things down.